doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+
+# Storage and Compute
+
+Materialize produces and maintains time-varying collections in two distinct ways:
+
+*   The **Storage** layer records explicit TVCs in response to direct instruction (e.g. user DML), integrations with external sources of data, and feedback from Materialize itself.
+    The representation in the Storage layer is as explicit update triples and maintained `since` and `upper` frontiers.
+    The primary property of the Storage layer is that, even across failures, it is able to reproduce all updates for times that lie between `since` and `upper` frontiers that it has advertised.
+
+*   The  **Compute** layer uses a view definition to transform input TVCs into the exactly corresponding output TVCs.
+    The output TVC varies *exactly* with the input TVCs:
+    For each `time`, the output collection at `time` equals the view applied to the input collections at `time`.
+
+The Storage and Compute layers both track `since` and `upper` frontiers for the collections they manage.
+For both layers, the `since` frontier is controlled by their users and it does not advance unless it is allowed.
+The `upper` frontier in Storage is advanced when it has sealed a frontier and made the data up to that point durable.
+The `upper` frontier in Compute is advanced when its inputs have advanced and its computation has pushed updates fully through the view definitions.
+
+## Capabilities
+
+The `since` and `upper` frontiers are allowed to advance through the use of "capabilities".
+
+* `ReadCapability(id, frontier)` prevents the `since` frontier associated with `id` from advancing beyond `frontier`.
+* `WriteCapability(id, frontier)` prevents the `upper` frontier associated with `id` from advancing beyond `frontier`.
+
+A capability presents the `id` and `frontier` values, but is identified by a system-assigned unique identifier.
+Any user presenting this identifier has the associated capability to read from or write to the associated pTVC.
+Actions which read or write will validate the capability (e.g. in case a lease has expired) and may return errors if the capability is not valid as stated.
+
+Capabilities are durable, and their state is tracked by the system who associates a unique identifier with each.
+Capabilities can be cloned and transfered to others, providing the same rights (and responsibilities).
+Capabilities can be downgraded to frontiers that are greater or equal to the held frontier, which may then allow the `since` or `upper` frontier to advance.
+Capabilities can be dropped, which releases the constraint on the frontier, and the whichever ability (read or write) the capability provided.
+These actions are mediated by the system, and take effect only when the system confirms the action.
+Users can defer and batch these actions, modulo lease timeouts, imposing only on the freshness and efficiency of the system.
+
+Capabilities may expire, as a lease, which may result in errors from attempts to exercise the capability.
```

Comment:
Yep, makes sense! Making leases optional at the caller’s discretion sounds great to me. Feels a bit like cooperative scheduling: we’ll need to note that acquiring a capability with no lease is a bit of an antisocial thing to do. And if it turns out to be a problem, we may need to reevaluate the API entirely.

Reply:
At least for a public API IMHO the lease should always be present (for the reasons stated by @frankmcsherry  above).

Reply:
> At least for a public API IMHO the lease should always be present (for the reasons stated by @frankmcsherry above).

I am pretty certain that there will be applications downstream of Materialize Platform that will require a read capability with no timeout for correctness. E.g., if you have a downstream application that 100% needs to see every single update in a collection. Yes, the risk is that you take down your cluster if your downstream application doesn't periodically downgrade the capability, but sometimes that's the tradeoff you want to make.

This is how PostgreSQL replication works, for example. The follower requests a replication slot that prevents WAL truncation; if the follower dies, the replication slot gets stuck, the WAL grows without bound, and the leader eventually runs out of disk and crashes. But it's the only way for the logical replication to be 100% correct.

Reply:
Minor nit, but PostgreSQL now has `max_slot_wal_keep_size` which allows you the admin to put bounds on the resource use of replication slots, to avoid folks taking it down.

I do agree there are cases we 100% don't want to time out a write capability: if we are producing the output to `SELECT COUNT(*) FROM ..` we don't want to time out and say "here is the correct answer" for something that is wrong. On the other hand, we eventually risk the same problem that PostgreSQL has for read capabilities (but perhaps we don't care for a while because cloud storage).

Anyhow, I think the current design of the parts puts most "critical" uses in and around the `Controller`, and if MZ is up and running we can have STORAGE, COMPUTE, and ADAPTER representatives live to forcibly downgrade capabilities, leaving only surprised uses of them (actual users, and maybe compute instances) need to pay attention to errors they might get back when using them.